src/expr/proof_generator.cpp

```diff
@@ -66,4 +68,94 @@ bool ProofGenerator::addProofTo(Node f, CDProof* pf, CDPOverwrite opolicy)
   return false;
 }
 
+void pfgEnsureClosed(Node proven,
+                     ProofGenerator* pg,
+                     const char* c,
+                     const char* ctx,
+                     bool reqGen)
+{
+  std::vector<Node> assumps;
+  pfgEnsureClosedWrt(proven, pg, assumps, c, ctx, reqGen);
+}
+
+void pfgEnsureClosedWrt(Node proven,
+                     ProofGenerator* pg,
+                     const std::vector<Node>& assumps,
+                     const char* c,
+                     const char* ctx,
+                     bool reqGen)
+{
+  if (!options::proofNew())
+  {
+    // proofs not enabled, do not do check
+    return;
+  }
+  bool isTraceDebug = Trace.isOn(c);
+  if (!options::proofNewEagerChecking() && !isTraceDebug)
+  {
+    // trace is off and proof new eager checking is off, do not do check
+    return;
+  }
```

Comment:
Do we need these hard coded tests here? It'd make more sense to me to test this in the caller.

Reply:
Yes we need them, or else this is a major pain to do over and over in the code, which is what I had before.  The caller should not be responsible for getting this behavior right.